src/main/java/org/assertj/core/internal/Iterables.java

```diff
@@ -339,10 +340,18 @@ public void assertHasSameSizeAs(AssertionInfo info, Iterable<?> actual, Iterable
    * @throws AssertionError if the given {@code Iterable} does not contain the given values.
    */
   public void assertContains(AssertionInfo info, Iterable<?> actual, Object[] values) {
-    final List<?> actualAsList = newArrayList(actual);
-    if (commonCheckThatIterableAssertionSucceeds(info, actualAsList, values)) return;
+    final Collection<?> actualAsCollection = ensureActualCanBeReadMultipleTimes(actual);
+    if (commonCheckThatIterableAssertionSucceeds(info, actualAsCollection, values)) return;
     // check for elements in values that are missing in actual.
-    assertIterableContainsGivenValues(actual.getClass(), actualAsList, values, info);
+    assertIterableContainsGivenValues(actual.getClass(), actualAsCollection, values, info);
+  }
+
+  private Collection<?> ensureActualCanBeReadMultipleTimes(Iterable<?> actual) {
+    if (actual instanceof Collection<?>) {
+      return (Collection<?>) actual; // Avoid copy when actual is already a Collection.
```

Comment:
There still may be edge cases like "user-provided-collection that loads items lazily from the database and allows calling .iterator() just once"

WDYT of adding an extra check like `actual.getClass().getClassLoader() == null` (or `==Collection.class.getClassLoader` for brevity)?